Wrong error message when an attribute is a substring of another

Component `Foo`:

```typescript
import Component from '@glint/environment-ember-loose/glimmer-component';

export interface IFoo {
  readonly Args: {
    readonly a: number;
    readonly aa: string;
  };
}

export default class Foo extends Component<IFoo> {}

declare module '@glint/environment-ember-loose/registry' {
  export default interface Registry {
    Foo: typeof Foo;
  }
}
```

Component `Bar`:

```typescript
import Component from '@glint/environment-ember-loose/glimmer-component';

export default class Bar extends Component {}
```

```handlebars
<Foo @aa="asd" @a="qwe" />
```

Leads to:

```
app/components/bar.hbs:1:7 - error TS2322: Type 'string' is not assignable to type 'number'.

1 <Foo @aa="asd" @a="qwe" />
        ~

  app/components/foo.ts:5:14
    5     readonly a: number;
                   ~
    The expected type comes from property 'a' which is declared here on type '{ readonly a: number; readonly aa: string; }'
```

Note the first squiggly line - it's on `aa` instead of on `a`.

@dfreeman - there are two more cases where this breaks:

1)

```hbs
{{foo aa="qwe" a="qwe"}}
```

```
app/bar.hbs:1:7 - error TS2322: Type 'string' is not assignable to type 'number'.

1 {{foo aa="qwe" a="qwe"}}
        ~

  app/foo.ts:5:14
    5     readonly a: number;
                   ~
    The expected type comes from property 'a' which is declared here on type '{ readonly a: number; readonly aa: string; }'
```

2)

```hbs
<Foo
  @aa={{@a}}
  @a="qwe" />
```

Where `bar` now has this definition:

```ts
export default class Bar extends Component<{ Args: { a: string } }> {}
```

```
app/bar.hbs:2:10 - error TS2322: Type 'string' is not assignable to type 'number'.

2   @aa={{@a}}
           ~

  app/foo.ts:5:14
    5     readonly a: number;
                   ~
    The expected type comes from property 'a' which is declared here on type '{ readonly a: number; readonly aa: string; }'
```

--- a/packages/transform/__tests__/offset-mapping.test.ts
+++ b/packages/transform/__tests__/offset-mapping.test.ts
@@ -370,6 +370,71 @@ describe('Source-to-source offset mapping', () => {
       });
     });
 
+    describe('location inference stress tests', () => {
+      test('matching in/out arg names', () => {
+        let { source, transformedModule } =
+          rewriteInlineTemplate({
+            contents: stripIndent`
+              <Foo
+                @foo={{@a}}
+                @a="bar" />
+            `,
+          }) ?? {};
+
+        let passedArgSourceOffset = source.contents.indexOf('{{@a}}') + '{{@'.length;
+        let passedArgTransformedOffset =
+          transformedModule.transformedContents.indexOf('args.a') + 'args.'.length;
+
+        expect(transformedModule.getOriginalOffset(passedArgTransformedOffset).offset).toBe(
+          passedArgSourceOffset
+        );
+
+        expect(transformedModule.getTransformedOffset(source.filename, passedArgSourceOffset)).toBe(
+          passedArgTransformedOffset
+        );
+
+        let argNameSourceOffset = source.contents.indexOf('@a=') + '@'.length;
+        let argNameTransformedOffset =
+          transformedModule.transformedContents.indexOf(', a:') + ', '.length;
+
+        expect(transformedModule.getOriginalOffset(argNameTransformedOffset).offset).toBe(
+          argNameSourceOffset
+        );
+
+        expect(transformedModule.getTransformedOffset(source.filename, argNameSourceOffset)).toBe(
+          argNameTransformedOffset
+        );
+      });
+
+      test('arg name subsequences', () => {
+        let { source, transformedModule } =
+          rewriteInlineTemplate({ contents: `{{foo aa="qwe" a="qwe"}}` }) ?? {};
+
+        let doubleASourceOFfset = source.contents.indexOf('aa=');
+        let doubleATransformedOffset = transformedModule.transformedContents.indexOf('aa:');
+
+        expect(transformedModule.getOriginalOffset(doubleATransformedOffset).offset).toBe(
+          doubleASourceOFfset
+        );
+
+        expect(transformedModule.getTransformedOffset(source.filename, doubleASourceOFfset)).toBe(
+          doubleATransformedOffset
+        );
+
+        let singleASourceOffset = source.contents.indexOf(' a=') + ' '.length;
+        let singleATransformedOFfset =
+          transformedModule.transformedContents.indexOf(' a:') + ' '.length;
+
+        expect(transformedModule.getOriginalOffset(singleATransformedOFfset).offset).toBe(
+          singleASourceOffset
+        );
+
+        expect(transformedModule.getTransformedOffset(source.filename, singleASourceOffset)).toBe(
+          singleATransformedOFfset
+        );
+      });
+    });
+
     test('Windows line endings', () => {
       let module = rewriteInlineTemplate({
         contents: `Hello, <World />!\r\n\r\n{{this.foo}}\r\n`,

--- a/packages/transform/src/template-to-typescript.ts
+++ b/packages/transform/src/template-to-typescript.ts
@@ -455,6 +455,8 @@ export function templateToTypescript(
             }
           });
 
+          start = rangeForNode(attr.value).end;
+
           if (index === dataAttrs.length - 1) {
             emit.text(' ');
           }
@@ -965,6 +967,8 @@ export function templateToTypescript(
         if (index === node.hash.pairs.length - 1) {
           emit.text(' ');
         }
+
+        start = rangeForNode(pair.value).end;
       }
 
       emit.text('}');

